actionpack/lib/action_controller/metal/request_forgery_protection.rb

```diff
@@ -366,9 +370,39 @@ def valid_per_form_csrf_token?(token, session) # :doc:
         end
       end
 
-      def real_csrf_token(session) # :doc:
-        session[:_csrf_token] ||= SecureRandom.base64(AUTHENTICITY_TOKEN_LENGTH)
-        Base64.strict_decode64(session[:_csrf_token])
+      # Returns a decoded CSRF token from the session or cookie and generates it
+      # if it wasn't set before.
+      def real_csrf_token(session)
+        # In order to smoothly switch between the cookie and the session modes,
+        # we need to provide backwards compatibility.
+        csrf_token = if persistent_csrf_token
+          cookies.encrypted[:_csrf_token] || session[:_csrf_token]
+        else
+          session[:_csrf_token] || cookies.encrypted[:_csrf_token]
```

Comment:
Exactly. Though it seems a rare case, it will be impossible to switch back without having this fallback.